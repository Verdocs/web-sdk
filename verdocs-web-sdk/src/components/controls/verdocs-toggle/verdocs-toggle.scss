@import '../../../theme.scss';

verdocs-toggle {
  font-family: $primary-font;

  .container {
    display: flex;
    flex-direction: row;
    flex-wrap: nowrap;
    background-color: $container-bg;
  }

  .label {
    font-size: 24px;
    font-weight: 700;
    color: $medium-bg;
    align-self: center;
    margin-right: 27px;
    font-family: sans-serif;
  }

  .buttons {
    display: flex;
    flex-direction: row;
    flex-wrap: nowrap;
  }

  .buttons button {
    width: 60px;
    height: 60px;
    border: 2px solid $item-border-color;
    border-radius: 5px;
    cursor: pointer;
  }

  .buttons button:not(:last-child) {
    margin-right: 11px;
  }

  .buttons svg {
    width: 30px;
    height: 30px;
  }

  .buttons button:hover {
    background-color: $item-border-color;
    transition: 0.25s;
  }

  .buttons button:hover svg {
    fill: $container-bg;
    transition: 0.25s;
  }

  .selected {
    background-color: $border-color;
    transition: 0.25s;
  }

  .selected svg {
    fill: #ffffff;
    transition: 0.25s;
  }
}
