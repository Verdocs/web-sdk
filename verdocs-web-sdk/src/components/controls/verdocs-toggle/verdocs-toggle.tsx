import {Component, Prop, h, State} from '@stencil/core';

export interface IButtons {
  id: string;
  icon: string;
}

export interface IToggleIconButtons {
  label: string;
  defaultSelection: number;
  buttons: IButtons[];
}

/**
 * Displays a label and a set of buttons, also allowing a default selection of a button
 */
@Component({
  tag: 'verdocs-toggle',
  styleUrl: 'verdocs-toggle.scss',
})
export class VerdocsToggle {
  /**
   * The tags to display
   */
  @Prop() options: IToggleIconButtons;

  /**
   * The "theme" to be used
   */
  @Prop() theme: 'light' | 'dark';

  @State() selectedOption: number = 0;

  componentWillLoad() {
    this.selectedOption = this.options?.defaultSelection || 0;
  }

  render() {
    return (
      <div class="container">
        <span class="label">{`${this.options.label}:`}</span>
        <div class="buttons">
          {this.options.buttons.map((button: IButtons, index: number) => (
            <button
              id={button.id}
              key={button.id}
              innerHTML={button.icon}
              class={index === this.selectedOption ? 'selected' : ''}
              onClick={() => {
                console.log('clicked', index);
                this.selectedOption = index;
              }}
            />
          ))}
        </div>
      </div>
    );
  }
}
