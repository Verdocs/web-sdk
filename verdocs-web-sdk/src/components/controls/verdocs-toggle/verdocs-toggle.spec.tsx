import { newSpecPage } from '@stencil/core/testing';
import { h } from '@stencil/core';
import { VerdocsToggle } from './verdocs-toggle';

describe('verdocs-toggle', () => {
  const options = {
    label: 'Toggle',
    defaultSelection: 0,
    buttons: [
      { id: 'btn1', icon: '<svg></svg>' },
      { id: 'btn2', icon: '<svg></svg>' }
    ]
  };

  it('renders with default props (with options)', async () => {
    const page = await newSpecPage({
      components: [VerdocsToggle],
      template: () => <verdocs-toggle options={options}></verdocs-toggle>,

    });
    expect(page.root).toMatchSnapshot();
  });

  it('renders with options and theme', async () => {
    const themedOptions = { ...options, defaultSelection: 1 };
    const page = await newSpecPage({
      components: [VerdocsToggle],
      template: () => <verdocs-toggle options={themedOptions} theme="dark"></verdocs-toggle>,
    });
    expect(page.root).toMatchSnapshot();
    expect(page.rootInstance.theme).toBe('dark');
    expect(page.rootInstance.options.label).toBe('Toggle');
  });

  it('changes selectedOption on button click', async () => {
    const page = await newSpecPage({
      components: [VerdocsToggle],
      template: () => <verdocs-toggle options={options}></verdocs-toggle>,
    });

    const buttons = page.root.querySelectorAll('button');
    expect(buttons.length).toBe(2);

    // Simulate click on the second button
    buttons[1].dispatchEvent(new Event('click', { bubbles: true }));
    expect(page.rootInstance.selectedOption).toBe(1);
  });
});
