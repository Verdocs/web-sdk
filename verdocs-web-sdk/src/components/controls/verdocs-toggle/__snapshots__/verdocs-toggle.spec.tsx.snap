// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verdocs-toggle renders with default props (with options) 1`] = `
<verdocs-toggle>
  <div class="container">
    <span class="label">
      Toggle:
    </span>
    <div class="buttons">
      <button class="selected" id="btn1">
        <svg></svg>
      </button>
      <button id="btn2">
        <svg></svg>
      </button>
    </div>
  </div>
</verdocs-toggle>
`;

exports[`verdocs-toggle renders with options and theme 1`] = `
<verdocs-toggle>
  <div class="container">
    <span class="label">
      Toggle:
    </span>
    <div class="buttons">
      <button id="btn1">
        <svg></svg>
      </button>
      <button class="selected" id="btn2">
        <svg></svg>
      </button>
    </div>
  </div>
</verdocs-toggle>
`;
